TypeScript semantic errors do not result in a failed source build.

While making some code generator changes I noticed the following:

```
[13:39:03] Starting 'Ice_objects:build'...
[13:39:05] Finished 'Ice_objects:build' after 1.84 s
[13:39:05] Starting 'Ice_operations:build'...
packages/test/Ice/operations/AMDMyDerivedClassI.ts(9,14): error TS2515: Non-abstract class 'AMDMyDerivedClassI' does not implement inherited abstract member opOneway from class 'MyDerivedClass'.
TypeScript: 1 semantic error
TypeScript: emit succeeded (with errors)
[13:39:07] Finished 'Ice_operations:build' after 2.51 s
```

--- a/js/gulpfile.js
+++ b/js/gulpfile.js
@@ -174,7 +174,31 @@ const tsCompilerOptions = {
     noImplicitReturns: true,
     noUnusedLocals: true,
     noUnusedParameters: true,
+    noEmitOnError: true,
 };
+
+// Compile TypeScript sources and fail the build if there are any errors.
+async function compileTypeScript(testDir) {
+    let errorCount = 0;
+    const reporter = {
+        error: (error) => {
+            console.error(error.message);
+            errorCount++;
+        },
+        finish: () => {},
+    };
+
+    await finished(
+        gulp
+            .src([`${testDir}/*.ts`, `!${testDir}/*.d.ts`])
+            .pipe(tsc(tsCompilerOptions, reporter))
+            .pipe(gulp.dest(testDir)),
+    );
+
+    if (errorCount > 0) {
+        throw new Error(`TypeScript: ${errorCount} error(s)`);
+    }
+}
 
 // Ice bundle: bundle the Ice runtime for browser usage.
 gulp.task("ice:bundle", async () => {
@@ -259,12 +283,7 @@ for (const name of tests) {
         }
 
         // Step 2: Compile .ts → .js in-place (for Node.js test runner)
-        await finished(
-            gulp
-                .src([`${testDir}/*.ts`, `!${testDir}/*.d.ts`])
-                .pipe(tsc(tsCompilerOptions))
-                .pipe(gulp.dest(testDir)),
-        );
+        await compileTypeScript(testDir);
 
         // Step 3: Bundle .js files for browser
         let input;

--- a/js/gulpfile.js
+++ b/js/gulpfile.js
@@ -181,7 +181,7 @@ const tsCompilerOptions = {
 async function compileTypeScript(testDir) {
     let errorCount = 0;
     const reporter = {
-        error: (error) => {
+        error: error => {
             console.error(error.message);
             errorCount++;
         },

